ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateProvider.java

```diff
@@ -73,149 +74,107 @@ public BonsaiWorldStateArchive(
             provider.createWorldStateStorage(DataStorageFormat.BONSAI),
         blockchain,
         Optional.empty(),
-        provider.isWorldStateSnappable(),
         cachedMerkleTrieLoader);
   }
 
-  public BonsaiWorldStateArchive(
+  public BonsaiWorldStateProvider(
       final BonsaiWorldStateKeyValueStorage worldStateStorage,
       final Blockchain blockchain,
       final Optional<Long> maxLayersToLoad,
       final CachedMerkleTrieLoader cachedMerkleTrieLoader) {
-    // overload while snapshots are an experimental option:
-    this(
-        worldStateStorage,
-        blockchain,
-        maxLayersToLoad,
-        DataStorageConfiguration.DEFAULT_BONSAI_USE_SNAPSHOTS,
-        cachedMerkleTrieLoader);
-  }
 
-  public BonsaiWorldStateArchive(
-      final BonsaiWorldStateKeyValueStorage worldStateStorage,
-      final Blockchain blockchain,
-      final Optional<Long> maxLayersToLoad,
-      final boolean useSnapshots,
-      final CachedMerkleTrieLoader cachedMerkleTrieLoader) {
-    this(
-        useSnapshots
-            ? new SnapshotTrieLogManager(
-                blockchain, worldStateStorage, maxLayersToLoad.orElse(RETAINED_LAYERS))
-            : new LayeredTrieLogManager(
-                blockchain, worldStateStorage, maxLayersToLoad.orElse(RETAINED_LAYERS)),
-        worldStateStorage,
-        blockchain,
-        useSnapshots,
-        cachedMerkleTrieLoader);
+    // TODO: de-dup constructors
+    this.trieLogManager =
+        new CachedWorldStorageManager(
+            this, blockchain, worldStateStorage, maxLayersToLoad.orElse(RETAINED_LAYERS));
+    this.blockchain = blockchain;
+    this.worldStateStorage = worldStateStorage;
+    this.persistedState = new BonsaiWorldState(this, worldStateStorage);
```

Comment:
Having a Java brain-fart moment...how can we reference `this` inside BonsaiWorldStateProvider constructor before BonsaiWorldStateProvider instance is fully initialised?

e.g. what if BonsaiWorldState constructor called `archive.getCachedMerkleTrieLoader()`?

Reply:
It is a theoretical risk, but not currently one because the archive is the only user of that object until the archive is fully constructed. 

However,  I have never been fond of having an archive reference in worldstate.  In a future PR IMO we could/should create functional interface and/or function references for the bits of the archive functionality that a worldstate needs so we can break the circular archive<->worldstate dependency.